File: answers/lintcode/Find Median from Data Stream.cpp
class MedianFinder {
public:

    // Adds a number into the data structure.
    void addNum(int num) {
        max_heap.push(num);
        min_heap.push(max_heap.top());
        max_heap.pop();
        while (min_heap.size() > max_heap.size() + 1) {
            int temp = min_heap.top();
            min_heap.pop();
            max_heap.push(temp);
        }
    }

    // Returns the median of current data stream
    double findMedian() {
        if (min_heap.size() > max_heap.size()) {
            return min_heap.top();
        }
        else {
            return (min_heap.top() + max_heap.top()) / 2.0;
        }
    }
    priority_queue<int> max_heap;
    priority_queue<int, vector<int>, greater<int> > min_heap;
};

// Your MedianFinder object will be instantiated and called as such:
// MedianFinder mf;
// mf.addNum(1);
// mf.findMedian();

File: answers/leetcode/Find Median from Data Stream/Find Median from Data Stream.java
class MedianFinder {
    
    private Queue<Integer> largeQueue;
    private Queue<Integer> smallQueue;

    /** initialize your data structure here. */
    public MedianFinder() {
        this.largeQueue = new PriorityQueue<>();
        this.smallQueue = new PriorityQueue<>(Collections.reverseOrder());
    }
    
    public void addNum(int num) {
        this.largeQueue.add(num);
        this.smallQueue.add(this.largeQueue.poll());
        while (this.smallQueue.size() > this.largeQueue.size() + 1) {
            this.largeQueue.add(this.smallQueue.poll());
        }
    }
    
    public double findMedian() {
        if (this.smallQueue.size() == this.largeQueue.size()) {
            return (this.smallQueue.peek() + this.largeQueue.peek()) / 2.0;
        } else {
            return this.smallQueue.peek();
        }
    }
}

/**
 * Your MedianFinder object will be instantiated and called as such:
 * MedianFinder obj = new MedianFinder();
 * obj.addNum(num);
 * double param_2 = obj.findMedian();
 */